### ISSUE TITLE
Fds leaking 

### ISSUE THREAD

**Issue by schnuffle** [matrix-org/synapse#1858]
So after gathering some facts, here's my case:

Synapse is runing fine for a while but then, as can be seen in the metrics some FDs are leaking.  I tried to gather as much info as I could, if anybody needs any other infos, just leave me a note.

Regards

schnuffle

# OS
- Centos 7
- Kernel 3.10.0-514.6.1.el7.x86_64
- Python python-2.7.5-48.el7.x86_64

# Python Env
```
appdirs==1.4.0
attrs==16.3.0
bleach==1.5.0
blist==1.3.6
canonicaljson==1.0.0
cffi==1.9.1
constantly==15.1.0
cryptography==1.7.2
daemonize==2.4.7
decorator==4.0.11
enum34==1.1.6
frozendict==1.2
html5lib==0.9999999
idna==2.2
incremental==16.10.1
ipaddress==1.0.18
Jinja2==2.9.5
ldap3==2.2.0
MarkupSafe==0.23
matrix-angular-sdk==0.6.8
matrix-synapse==0.18.7
matrix-synapse-ldap3==0.1.2
msgpack-python==0.4.8
netaddr==0.7.19
olefile==0.44
packaging==16.8
Paste==2.0.3
Pillow==4.0.0
psutil==5.1.0
py-bcrypt==0.4
pyasn1==0.1.9
pyasn1-modules==0.0.8
pycparser==2.17
pycrypto==2.6.1
pydenticon==0.3
pymacaroons-pynacl==0.9.3
PyNaCl==0.3.0
pyOpenSSL==16.2.0
pyparsing==2.1.10
pysaml2==3.0.2
python-dateutil==2.6.0
pytz==2016.10
PyYAML==3.12
repoze.who==2.3
requests==2.13.0
service-identity==16.0.0
signedjson==1.0.0
simplejson==3.10.0
six==1.10.0
Twisted==16.6.0
ujson==1.35
unpaddedbase64==1.1.0
WebOb==1.7.1
zope.interface==4.3.3
```
# Metrics
![image](https://cloud.githubusercontent.com/assets/5358397/22349896/f252a8b6-e411-11e6-96f9-872132392408.png)

# Lsof
[lsof.txt](https://github.com/matrix-org/synapse/files/733594/lsof.txt)

# Log excerpt
Attention file is 9MB
[synapse.log-2017012612.gz](https://github.com/matrix-org/synapse/files/733604/synapse.log-2017012612.gz)

```
Jan 26 03:34:12 kolab python2.7: exec code in run_globals
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/synapse/app/homeserver.py", line 491, in <module>
Jan 26 03:34:12 kolab python2.7: main()
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/synapse/app/homeserver.py", line 487, in main
Jan 26 03:34:12 kolab python2.7: run(hs)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/synapse/app/homeserver.py", line 479, in run
Jan 26 03:34:12 kolab python2.7: in_thread()
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/synapse/app/homeserver.py", line 461, in in_thread
Jan 26 03:34:12 kolab python2.7: reactor.run()
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/internet/base.py", line 1199, in run
Jan 26 03:34:12 kolab python2.7: self.mainLoop()
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/internet/base.py", line 1211, in mainLoop
Jan 26 03:34:12 kolab python2.7: self.doIteration(t)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/internet/epollreactor.py", line 235, in doPoll
Jan 26 03:34:12 kolab python2.7: log.callWithLogger(selectable, _drdw, selectable, fd, event)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/python/log.py", line 103, in callWithLogger
Jan 26 03:34:12 kolab python2.7: return callWithContext({"system": lp}, func, *args, **kw)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/python/log.py", line 86, in callWithContext
Jan 26 03:34:12 kolab python2.7: return context.call({ILogContext: newCtx}, func, *args, **kw)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
Jan 26 03:34:12 kolab python2.7: return self.currentContext().callWithContext(ctx, func, *args, **kw)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
Jan 26 03:34:12 kolab python2.7: return func(*args,**kw)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 610, in _doReadOrWrite
Jan 26 03:34:12 kolab python2.7: self._disconnectSelectable(selectable, why, inRead)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 255, in _disconnectSelectable
Jan 26 03:34:12 kolab python2.7: selectable.connectionLost(f)
Jan 26 03:34:12 kolab python2.7: File "/home/synapse/pyenv/lib/python2.7/site-packages/twisted/internet/tcp.py", line 350, in connectionLost
Jan 26 03:34:12 kolab python2.7: tb=''.join(traceback.format_stack())
Jan 26 03:34:12 kolab python2.7: 2017-01-26 03:34:12,651 - twisted - 131 - INFO - - Connection.connectionLost | fileno: -1 | addrs: local: '?', remote: '?'
Jan 26 03:34:12 kolab python2.7: File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
Jan 26 03:34:12 kolab python2.7: "__main__", fname, loader, pkg_name)
Jan 26 03:34:12 kolab python2.7: File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
```
# Install procedure
```
virtualenv-2.7 --no-site-packages pyenv
. pyenv/bin/activate
mkdir src ;; cd src
git clone https://github.com/matrix-org/synapse.git
cd synapse
git checkout develop
python synapse/python_dependencies.py | xargs -n1 pip install --upgrade
pip install --upgrade .
```

Finally I introduced @markwilliams [Patch](https://gist.github.com/markrwilliams/5bff78dece0f9724f2ddad4934a87c2d)

# Systemd 
```
[Unit]
Description=Synapse Matrix homeserver

[Service]
Type=simple
StandardOutput=syslog
SyslogFacility=local5
User=synapse
Group=synapse
#EnvironmentFile=-/etc/sysconfig/synapse
WorkingDirectory=/home/synapse
ExecStart=/home/synapse/pyenv/bin/python2.7 -m synapse.app.homeserver --config-path=/home/synapse/etc/schnuffle.de.yaml --log-config=/home/synapse/etc/schnuffle.de.log.config

[Install]
WantedBy=multi-user.target
```
# Config
```
tls_certificate_path: "/etc/letsencrypt/live/kolab.schnuffle.de/fullchain.pem"
tls_dh_params_path: "/home/synapse/etc/schnuffle.de.tls.dh"
no_tls: False
tls_fingerprints: []
server_name: "schnuffle.de"
pid_file: /home/synapse/db/schnuffle.pid
web_client: False
public_baseurl: https://kolab.schnuffle.de:8448/
soft_file_limit: 0
listeners:
  - port: 8448
    bind_address: ''
    type: http
    tls: true
    # Use the X-Forwarded-For (XFF) header as the client IP and not the
    # actual client IP.
    x_forwarded: false
    resources:
      - names:
          - client     # The client-server APIs, both v1 and v2
        #  - webclient  # The bundled webclient.
        compress: true
      - names: [federation]  # Federation APIs
        compress: false
  - port: 8008
    tls: false
    bind_address: ''
    type: http
    x_forwarded: false
    resources:
      - names: [client, webclient]
        compress: true
      - names: [federation]
        compress: false
  # Turn on the twisted ssh manhole service on localhost on the given
  # port.
  # - port: 9000
  #   bind_address: 127.0.0.1
  #   type: manhole
database:
  name: "sqlite3"
  args:
    database: "/home/synapse/db/schnuffle.db"
event_cache_size: "10K"
verbose: 0
log_config: "/home/synapse/etc/schnuffle.de.log.config"
rc_messages_per_second: 0.2
rc_message_burst_count: 10.0
federation_rc_window_size: 1000
federation_rc_sleep_limit: 10
federation_rc_sleep_delay: 500
federation_rc_reject_limit: 50
federation_rc_concurrent: 3
media_store_path: "/home/synapse/media_store"
uploads_path: "/home/synapse/uploads"
max_upload_size: "10M"
max_image_pixels: "32M"
dynamic_thumbnails: false
thumbnail_sizes:
- width: 32
  height: 32
  method: crop
- width: 96
  height: 96
  method: crop
- width: 320
  height: 240
  method: scale
- width: 640
  height: 480
  method: scale
- width: 800
  height: 600
  method: scale
max_spider_size: "10M"
enable_registration_captcha: False
recaptcha_siteverify_api: "https://www.google.com/recaptcha/api/siteverify"
turn_uris: ["turn:kolab.schnuffle.de:3478?transport=udp","turn:kolab.schnuffle.de:3478?transport=tcp"]
turn_user_lifetime: "1h"
enable_registration: False
bcrypt_rounds: 12
allow_guest_access: False
trusted_third_party_id_servers:
    - matrix.org
    - vector.im
enable_metrics: True
metrics_port: 9091
report_stats: False
room_invite_state_types:
    - "m.room.join_rules"
    - "m.room.canonical_alias"
    - "m.room.avatar"
    - "m.room.name"
app_service_config_files: []
expire_access_token: False
perspectives:
  servers:
    "matrix.org":
        "ed25519:auto":
password_config:
   enabled: true
   # DO NOT CHANGE THIS AFTER INITIAL SETUP!
```
# Log Config
```

version: 1

formatters:
  precise:
   format: '%(asctime)s - %(name)s - %(lineno)d - %(levelname)s - %(request)s- %(message)s'

filters:
  context:
    (): synapse.util.logcontext.LoggingContextFilter
    request: ""

handlers:
  console:
    class: logging.StreamHandler
    formatter: precise
    filters: [context]

loggers:
    synapse:
        level: DEBUG

    synapse.storage.SQL:
        level: DEBUG

root:
    level: DEBUG
    handlers: [console]
```

**Comment by schnuffle** [comment]
I added another patch to see if that brings some progress: [Twisted commit](https://github.com/twisted/twisted/commit/5f37cd1b83a2609f23a9dab46fd023cc941153f2)

Result, the service is still leaking

**Comment by schnuffle** [comment]
After comparing a pip freeze with @seequ:matrix.org pip freeze I saw some version differences in the modules. Now I reinstalled all modules from seequs requireements.

```
appdirs==1.4.0
attrs==16.3.0
bleach==1.5.0
blist==1.3.6
canonicaljson==1.0.0
cffi==1.9.1
constantly==15.1.0
cryptography==1.7.1
daemonize==2.4.7
decorator==4.0.11
enum34==1.1.6
frozendict==1.2
html5lib==0.9999999
idna==2.2
incremental==16.10.1
ipaddress==1.0.17
Jinja2==2.9.4
ldap3==2.1.1
lxml==3.7.1
MarkupSafe==0.23
matrix-angular-sdk==0.6.8
matrix-synapse==0.18.7
matrix-synapse-ldap3==0.1.1
msgpack-python==0.4.8
MySQL-python==1.2.5
netaddr==0.7.19
olefile==0.44
packaging==16.8
Paste==2.0.3
Pillow==4.0.0
psutil==5.0.1
py-bcrypt==0.4
pyasn1==0.1.9
pyasn1-modules==0.0.8
pycparser==2.17
pycrypto==2.6.1
pydenticon==0.3
pymacaroons-pynacl==0.9.3
PyNaCl==0.3.0
pyOpenSSL==16.2.0
pyparsing==2.1.10
pysaml2==3.0.2
python-dateutil==2.6.0
pytz==2016.10
PyYAML==3.12
repoze.who==2.3
requests==2.12.5
service-identity==16.0.0
signedjson==1.0.0
simplejson==3.10.0
six==1.10.0
Twisted==16.6.0
ujson==1.35
unpaddedbase64==1.1.0
WebOb==1.7.0
zope.interface==4.3.3
```

I'll see if that changes something

**Comment by schnuffle** [comment]
Okay, synapse is now runing long enough without leaking that I would say, the problems seems to be resolved.
When i compare the difference, the leak was introduced by one of the two  module differences:

# Module versions with leaking problems
- ipaddress: 1.0.18
- lxml: 3.7.2
# Module versions without leaking problem
- ipaddress: 1.0.17
- lxml: 3.7.1

My quest is finished for the moment. In case someone is intrested in going deeper to the problem, I'm willing to help
-

**Comment by schnuffle** [comment]
I finally upgraded to 0.19.0 and right the server is going strong for 48h already

**Comment by schnuffle** [comment]
Okay, the reason for the leaks is not about the modules :).

I close that issue and will open a new one as soon as my new idea has proofs

**Comment by ara4n** [comment]
reopening this given @schnuffle's FDs are still leaking afaik.

**Comment by ara4n** [comment]
see also #1778

**Comment by richvdh** [comment]
@schnuffle's leaked connections show the following in lsof:

```
python2.7 12749 synapse   39u     sock                0,7       0t0   2131955 protocol: TCP
```

I don't really know what that means.

**Comment by richvdh** [comment]
Turns out that `sw1v.org` leaks FDs too.

I left a strace running for a while, and here is an example of an FD being opened and leaked:

```
9:50:58 epoll_wait(12, [{EPOLLIN, {u32=49, u64=9386145937253990449}}], 42, 171) = 1
09:50:58 accept(5, {sa_family=AF_INET, sin_port=htons(42796), sin_addr=inet_addr("5.9.168.181")}, [16]) = 71
09:50:58 fcntl(71, F_GETFD)             = 0
09:50:58 fcntl(71, F_SETFD, FD_CLOEXEC) = 0
09:50:58 fcntl(71, F_GETFL)             = 0x2 (flags O_RDWR)
09:50:58 fcntl(71, F_SETFL, O_RDWR|O_NONBLOCK) = 0
09:50:58 epoll_ctl(12, EPOLL_CTL_ADD, 71, {EPOLLIN, {u32=71, u64=9386145937253990471}}) = 0
09:50:58 epoll_wait(12, [{EPOLLIN, {u32=71, u64=9386145937253990471}}], 42, 36) = 1
09:50:58 recvfrom(71, "\26\3\1\0\317\1\0\0\313\3\3\201a,\230+\237|[\263b3\362;\330\270J\2Lf\326\245y[\205\211\361\327\305W\234\0$\0\0:\0\243\0\237\0k\0j\0009\0008\0\210\0\207\0\242\0\236\0g\0@\0003\0002\0E\0D\3000\300,\300(\300$\300\24\300\n\300/\300+\300'\300#\300\23\300\t\0\377\1\0\0h\0\v\0\4\3\0\1\2\0\n\0\34\0\32
09:50:58 epoll_ctl(12, EPOLL_CTL_MOD, 71, {EPOLLIN|EPOLLOUT, {u32=71, u64=9386145937253990471}}) = 0
09:50:58 epoll_wait(12, [{EPOLLOUT, {u32=71, u64=9386145937253990471}}], 42, 26) = 1
09:50:58 sendto(71, "\26\3\3\0I\2\0\0E\3\3\200t$k\253z\216\224\v\1\2PK\213\35\334\364\344E'\311\22\355m\231\300yAz\271\21\351\0\0\237\0\0\35\377\1\0\1\0\0#\0\0\0\17\0\1\1\0\20\0\v\0\t\10http/1.1\26\3\3\2\253\v\0\2\247\0\2\244\0\2\2410\202\2\2350\202\1\205\2\2\3\3500\r\6\t*\206H\206\367\r\1\1\v\5\0000\0241\0220\20\6\
09:50:58 epoll_ctl(12, EPOLL_CTL_MOD, 71, {EPOLLIN, {u32=71, u64=9386145937253990471}}) = 0
09:50:58 epoll_wait(12, [{EPOLLIN, {u32=71, u64=9386145937253990471}}], 42, 0) = 1
09:50:58 recvfrom(71, "\26\3\3\1\6\20\0\1\2\1\0\273\311\r\352%\371\361\10\221\220\226*\311<\300\3635i\230\6\241\361\2561\312dtj\376\275\25\251<\311\217\324\331\\\240 o\225f^!\vd\vJ\23|\320w$\264\337\346\261[JM0)k\356\225W\17\203\300\324\265\331\325e>G\1L\5b+\222\0375\300\264\317R\325\1r\325\340:j\377\301\307\325%\31
09:50:58 epoll_ctl(12, EPOLL_CTL_MOD, 71, {EPOLLIN|EPOLLOUT, {u32=71, u64=9386145937253990471}}) = 0
09:50:58 epoll_wait(12, [{EPOLLOUT, {u32=71, u64=9386145937253990471}}], 42, 11) = 1
09:50:58 sendto(71, "\26\3\3\0\252\4\0\0\246\0\0\1,\0\2400\205\223~\32\374\242\200\36Y\26h\254XT\370\337O\236\333]_?!\20\236\324\360\2565\353\3350\3709 \230[\376\370\354\337\214\247\230\323\352\304rD\16G\341\311\267\260\274\307\34\227\204&\251 \333\372\310y\274\304}\224,\7\n\213R(\305\366\2106v\345X\243\266Z\237\343
09:50:58 epoll_ctl(12, EPOLL_CTL_MOD, 71, {EPOLLIN, {u32=71, u64=9386145937253990471}}) = 0
09:50:58 epoll_wait(12, [{EPOLLIN, {u32=71, u64=9386145937253990471}}], 42, 499) = 1
09:50:58 recvfrom(71, "\27\3\3\0N\222\322\217T(\252R\315\266\315\353\3538u\355\366\257\202\205\32#\214E\331z\352\307%\337\216\237^\323\314,\317\34:r&l\215\35\220\f\355\300\0(A\246\203\204\177\16)\337\372\2410\270m\223\320Rm\30\315M\361\270\263e\220\351\224\240_\27\3\3\0(\222\322\217T(\252R\316\371l\364U6rJ\204\256\3
09:50:58 epoll_ctl(12, EPOLL_CTL_DEL, 71, 0x7fffa30b5800) = 0
09:50:58 getsockname(71, {sa_family=AF_INET, sin_port=htons(8448), sin_addr=inet_addr("109.74.200.146")}, [16]) = 0
09:50:58 epoll_ctl(12, EPOLL_CTL_ADD, 71, {EPOLLOUT, {u32=71, u64=9386145937253990471}}) = 0
09:50:58 epoll_wait(12, [{EPOLLOUT, {u32=71, u64=9386145937253990471}}], 42, 491) = 1
09:50:58 sendto(71, "\27\3\3\0\270O\333\327\204\276i\30\1kt_A\7\5\244\30g\356\377X\355\2365\246C\226\20VY\374\2543\317\357\337{\361\226\216\216\306P\177\311\301?\177T`~\302\254(CB-\275\2212e\323(\355\254\357\356\17f\342\365\256\340\203YZ\t\257b\373\v\311\365M\346\277SaR\221\n\303sT?\361\tLh\377\217\266\350\332\367\2
09:50:58 epoll_ctl(12, EPOLL_CTL_DEL, 71, 0x7fffa30b6db0) = 0
```

In other words:
* it was an inbound federation request (from 5.9.168.181)
* we read three packets from it, and sent replies, but then forgot about it.
* we don't register any interest in reading further packets from it - or noting when it gets closed. Hence the leak.

There is nothing in the logs about it.

It looks like we have completed TLS negotiation on it, but haven't then read any HTTP requests from it. Perhaps something went wrong in the TLS layer?

Next step is probably to try to get some more logging out of the accept/TLS code to see what it thinks is going on.

**Comment by schnuffle** [comment]
@richvdh asked me for a log excerpt, here it is.

It's right before a restart and the following hour, where synapse started leaking from the beginning. Here are the corresponding metrics:

![logmetrics](https://cloud.githubusercontent.com/assets/5358397/23805408/0d4a6f64-05be-11e7-8331-78d89159aefa.png)

Here's the compressed Log excerpt, it's 11MB compressed:
It starts right before the restart and then an additional hour till 13:15h.

[logexcerpt.txt.gz](https://github.com/matrix-org/synapse/files/834635/logexcerpt.txt.gz)

**Comment by richvdh** [comment]
So, it turns out that I was seeing a completely different issue to @schnuffle: I was still on Twisted 16.3, which means that I was probably suffering from https://twistedmatrix.com/trac/ticket/8992. (I have now upgraded to Twisted 17.1, and no leaks so far).

Schnuffle's logs show incoming requests for some media at `kolab.schnuffle.de` (the media in question appears to be an avatar for an old user).   `kolab.schnuffle.de` doesn't match the `server_name`, so the server forwards the request on... straight back to itself.

Surely we shouldn't forward requests for remote media that come in over federation?

**Comment by schnuffle** [comment]
when I first setup the instance it was under the domain kolab.schnuffle.de. After reading about the DNS SRV setup, i switched to schnuffle.de and rerstarted with new keys and db.

**Comment by richvdh** [comment]
Right; I've filed a new bug #1991 to track the problem you are seeing.

**Comment by richvdh** [comment]
schnuffle reports his woes seem to have been solved by the fix to #1991
